### PR TITLE
Preprocessing pipeline

### PR DESCRIPTION
**Pull request by jrx-napoli**
🙏

### CODE REVIEW

File: configs/preprocessing.yaml [hunk 1]
```diff
@@ -1,42 +1,43 @@
...
     margin: 0.2  # Margin around detected face as percentage of face size
 
   image_processing:
-    target_size: [224, 224]
+    target_size: [ 224, 224 ]
     normalization:
-      mean: [0.485, 0.456, 0.406]  # ImageNet mean
-      std: [0.229, 0.224, 0.225]   # ImageNet std
+      mean: [ 0.485, 0.456, 0.406 ]  # ImageNet mean
+      std: [ 0.229, 0.224, 0.225 ]   # ImageNet std
```
Comment (matikurcze00) [inline configs/preprocessing.yaml]: Co oznaqcza normalization w tym wypadku?

File: configs/preprocessing.yaml [hunk 2]
```diff
@@ -1,42 +1,43 @@
...
     augmentations:
```
Comment (matikurcze00) [inline configs/preprocessing.yaml]: nie powinniśmy od razu dodać parametru % augmentacji?

File: configs/preprocessing.yaml [hunk 3]
```diff
@@ -1,42 +1,43 @@
...
       enabled: true
       horizontal_flip: true
       rotation_range: 15
-      brightness_range: [0.8, 1.2]
-      contrast_range: [0.8, 1.2]
+      brightness_range: [ 0.8, 1.2 ]
+      contrast_range: [ 0.8, 1.2 ]
 
   audio_processing:
```
Comment (matikurcze00) [inline configs/preprocessing.yaml:19]: Nie chcemy dodać augmentacji w audio?

File: configs/preprocessing.yaml [hunk 4]
```diff
@@ -1,42 +1,43 @@
...
+    enabled: true
```
Comment (matikurcze00) [inline configs/preprocessing.yaml:20]: Parametr też powinien być w video?

File: configs/preprocessing.yaml [hunk 5]
```diff
@@ -1,42 +1,43 @@
...
-    n_mels: 80
-    n_fft: 2048
-    hop_length: 512
+    n_mels: 64
+    n_fft: 1024
+    hop_length: 256
 
 output:
-  format: "numpy"  # Options: "hdf5", "numpy", "torch"
   batch_size: 100
```
Comment (matikurcze00) [inline configs/preprocessing.yaml]: Chyba nie używamy tego parametru nigdzie

File: configs/preprocessing.yaml [hunk 6]
```diff
@@ -1,42 +1,43 @@
...
-  compression: "gzip"
-  compression_level: 9 
\ No newline at end of file
+
+  # WebDataset configuration
+  webdataset:
+    prefix: "shard"
+    max_shard_size_mb: 1024
+    image_codec: "webp"     # "webp" or "jpeg"
+    image_quality: 90        # 1..100
```
Comment (matikurcze00) [inline configs/preprocessing.yaml]: #1-100

File: docs/webdataset.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    shuffle=True,
+    num_workers=4
+)
+
+# Training loop
+for batch in train_loader:
+    frames = batch["data"]      # [B, T, H, W, 3]
+    labels = batch["label"]     # [B]
+    # ... training code
+```
```
Comment (matikurcze00) [inline docs/webdataset.md]: powinniśmy zostawiać pustą linię na końcu plików .md

File: src/data/shard_writer.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        total = 0
+        for i in range(frames_rgb.shape[0]):
+            data, ext = self._encode_frame(frames_rgb[i])
+            buffers.append(data)
+            total += len(data)
+        meta_bytes = json.dumps(metadata).encode("utf-8")
+        total += len(meta_bytes)
+        return total, buffers, ext or ".jpg"
+
+    def add_sample(
```
Comment (matikurcze00) [inline src/data/shard_writer.py]: rozbiłbym tę funkcję na parę podfunkcji, bo ma 80 linijek

File: src/data/shard_dataset.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        entries: List[Tuple[str, str, str, int, int, str]] = []
+        with open(path, "r", encoding="utf-8") as f:
+            # skip header
+            next(f)
+            for line in f:
+                sample_id, shard, sample_dir, num_frames, label, meta_json = line.rstrip("\n").split(",", 5)
+                entries.append((sample_id, shard, sample_dir, int(num_frames), int(label), meta_json))
+        return entries
+
+    def _iter_samples(self):
```
Comment (matikurcze00) [inline src/data/shard_dataset.py]: rozbiłbym tę funkcję bo jest długa i ma 4 pętle w sobie

File: src/data/shard_dataset.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                    for i in range(num_frames):
+                        # Try webp then jpg
+                        for ext in (".webp", ".jpg"):
+                            name = f"{sample_dir}/frame_{i:03d}{ext}"
+                            mem = members.get(name)
+                            if mem is not None:
+                                buf = tar.extractfile(mem).read()
+                                frames.append(_decode_image(buf))
+                                break
+                        else:
```
Comment (matikurcze00) [inline src/data/shard_dataset.py]: else nie jest przesunięty za bardzo wzgędem IFa?

File: src/data/fakeavceleb_preprocessor.py [hunk 11]
```diff
@@ -132,115 +131,24 @@ def process_dataset(self, input_dir: str, output_dir: str):
...
                 except Exception as e:
+                    # Log and continue with next video
                     print(f"\nError processing {video_path}: {str(e)}")
                     continue
 
         # Finalize output storage
-        self._finalize_output_storage(output_dir, output_format)
+        self._finalize_output_storage(output_dir)
 
         # Save dataset statistics
```
Comment (matikurcze00) [inline src/data/fakeavceleb_preprocessor.py]: niepotrzebne komentarze

File: src/data/base_preprocessor.py [hunk 12]
```diff
@@ -1,267 +1,413 @@
...
+            start_frame: Starting frame index in the video
+            clip_len: Number of frames in the clip
+            fps: Frames per second of the video
+            mel_per_second: Mel frames per second
+
+        Returns:
+            Frame-level mel spectrograms [clip_len, n_mels, mel_frames_per_frame]
+        """
+        n_mels = audio_mel_full.shape[0]
+        mel_frames_per_frame = max(1, int(round(mel_per_second / fps)))
```
Comment (matikurcze00) [inline src/data/base_preprocessor.py:115]: nie powinniśmy zastosować overlappingu? Lepiej by mel_frames_per_frame było większe (np. 2 razy większe) i trzeba wtedy uwazać na pierwsze i ostatnie spektogramy (będą takie przed drugie/przedostatnie)

File: src/data/base_preprocessor.py [hunk 13]
```diff
@@ -1,267 +1,413 @@
...
+                return None
+            mel_kwargs = {"y": y, "sr": sr, "n_mels": n_mels, "n_fft": n_fft, "hop_length": hop_length}
+            S = librosa.feature.melspectrogram(**mel_kwargs)
+            S_dB = librosa.power_to_db(S, ref=np.max)
+            return S_dB.astype(np.float32)
+        except Exception:
+            return None
+
+    @staticmethod
+    def extract_frame_level_mel(
```
Comment (matikurcze00) [inline src/data/base_preprocessor.py:93]: Zamykamy się tylko na mel? Pewnie łatwo jest zrobić też spektogramy, więc można zmienić nazwę na "extract_frame_level_spect" i trzymać się nazwy spektogram (skoro mel to też spektogram)

**Review by matikurcze00 — COMMENTED** [review]
(y)

File: configs/preprocessing.yaml (hunk 1 above)
Comment (jrx-napoli) [reply configs/preprocessing.yaml]: standardowo, skaluje wartości pixeli

File: configs/preprocessing.yaml (hunk 2 above)
Comment (jrx-napoli) [reply configs/preprocessing.yaml]: to jest nieuzywana sekcja configu, pozostalosc z czasow prototypu. usuwam

File: configs/preprocessing.yaml (hunk 3 above)
Comment (jrx-napoli) [reply configs/preprocessing.yaml:19]: Nawet jeśli to nie wiem w tym momencie jakie, myślę że do przemyślenia po jakichkolwiek wstępnych eksperymentach

File: configs/preprocessing.yaml (hunk 4 above)
Comment (jrx-napoli) [reply configs/preprocessing.yaml:20]: a planujemy pracować z jakimkolwiek datasetem/metodą fully audio?

File: configs/preprocessing.yaml (hunk 5 above)
Comment (jrx-napoli) [reply configs/preprocessing.yaml]: fakty, zostało z poprzedniej wersji

File: src/data/base_preprocessor.py (hunk 13 above)
Comment (jrx-napoli) [reply src/data/base_preprocessor.py:93]: Podejrzewam że będziemy używali tylko mel, z tego co wiem daje na ogół najlepsze wyniki (i ogólnie wątpię żebyśmy mieli dość zasobów żeby testować coś innego). Tak więc chyba bym zostawił, w tej metodzie chodzi obecnie dokładnie o to.

File: src/data/base_preprocessor.py (hunk 12 above)
Comment (jrx-napoli) [reply src/data/base_preprocessor.py:115]: chyba nie do końca rozumiem - wyciagam tylko fragmety spektrogramu dopasowane do konkretnej klatki. Co ma się overlapować? Mamy już overlap samych clipów

File: src/data/fakeavceleb_preprocessor.py (hunk 11 above)
Comment (jrx-napoli) [reply src/data/fakeavceleb_preprocessor.py]: kolega zostawił

File: src/data/shard_writer.py (hunk 8 above)
Comment (jrx-napoli) [reply src/data/shard_writer.py]: you got it

File: src/data/shard_dataset.py (hunk 10 above)
Comment (jrx-napoli) [reply src/data/shard_dataset.py]: to jest else do for-a, nie ifa ;p

File: src/data/shard_dataset.py (hunk 9 above)
Comment (jrx-napoli) [reply src/data/shard_dataset.py]: done

File: src/data/base_preprocessor.py (hunk 12 above)
Comment (matikurcze00) [reply src/data/base_preprocessor.py:115]: na WIMU było mówione, że przy przetwarzaniu audio spektogramy nachodzą na siebie, bo to zapewnia ciagłość sygnału (żadna część sygnału audio nie zostanie pominięta) oraz wygładzenie

**Comment by matikurcze00** [comment]
(y)

File: src/data/base_preprocessor.py [hunk 14]
```diff
@@ -95,49 +95,62 @@ def extract_frame_level_mel(
...
+            
+            # Add overlap by extending the window backwards and forwards
+            mel_start_with_overlap = mel_start - overlap_frames
+            mel_end_with_overlap = mel_start + mel_frames_per_frame + overlap_frames
 
             # Ensure bounds
-            mel_start = max(0, mel_start)
-            mel_end = min(audio_mel_full.shape[1], mel_end)
+            mel_start_with_overlap = max(0, mel_start_with_overlap)
+            mel_end_with_overlap = min(audio_mel_full.shape[1], mel_end_with_overlap)
```
Comment (matikurcze00) [inline src/data/base_preprocessor.py:140]: Pierwsza i ostatnia klatka będą miały inną długość. Zrobiłbym dla nich oddzielnego if'a by miały tę samą długość by na pewno nic nie zepsuły, albo trzeba chociaż dodać zerowy padding

**Review by matikurcze00 — APPROVED** [review]

File: src/data/base_preprocessor.py (hunk 14 above)
Comment (jrx-napoli) [reply src/data/base_preprocessor.py:140]: To już jest obsłużone, zaraz poniżej (linie 147 - 153)
